pkg/koordlet/metricsadvisor/be_collector.go

```diff
@@ -0,0 +1,140 @@
+/*
+ Copyright 2022 The Koordinator Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package metricsadvisor
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+func (c *collector) collectBECPUResourceMetric() {
+	klog.V(6).Info("collectBECPUResourceMetric start")
+
+	realMilliLimit, err := getBECPURealMilliLimit()
+	if err != nil {
+		klog.Errorf("getBECPURealMilliLimit failed, error: %v", err)
+		return
+	}
+
+	beCPURequest := c.getBECPURequestSum()
+
+	beCPUUsageCores, err := c.getBECPUUsageCores()
+	if err != nil {
+		klog.Errorf("getBECPUUsageCores failed, error: %v", err)
+		return
+	}
+
+	if beCPUUsageCores == nil {
+		klog.Info("beCPUUsageCores is nil")
+		return
+	}
+
+	beCPUMetric := metriccache.BECPUResourceMetric{
+		CPUUsed:      *beCPUUsageCores,
+		CPURealLimit: *resource.NewMilliQuantity(int64(realMilliLimit), resource.DecimalSI),
+		CPURequest:   beCPURequest,
+	}
+
+	collectTime := time.Now()
+	err = c.metricCache.InsertBECPUResourceMetric(collectTime, &beCPUMetric)
+	if err != nil {
+		klog.Errorf("InsertBECPUResourceMetric failed, error: %v", err)
+		return
+	}
+	klog.V(6).Info("collectBECPUResourceMetric finished")
+}
+
+func getBECPURealMilliLimit() (int, error) {
+	limit := 0
+
+	cpuSet, err := util.GetRootCgroupCurCPUSet(corev1.PodQOSBestEffort)
+	if err != nil {
+		return 0, err
+	}
+	limit = len(cpuSet) * 1000
+
+	cfsQuota, err := util.GetRootCgroupCurCFQuota(corev1.PodQOSBestEffort)
+	if err != nil {
+		return 0, err
+	}
+
+	// -1 means not suppress by cfs_quota
+	if cfsQuota == -1 {
+		return limit, nil
+	}
+
+	cfsPeriod, err := util.GetRootCgroupCurCFSPeriod(corev1.PodQOSBestEffort)
+	if err != nil {
+		return 0, err
+	}
+
+	limitByCfsQuota := int(cfsQuota * 1000 / cfsPeriod)
+
+	if limitByCfsQuota < limit {
+		limit = limitByCfsQuota
+	}
+
+	return limit, nil
+}
+
+func (c *collector) getBECPURequestSum() resource.Quantity {
+	requestSum := int64(0)
+	for _, podMeta := range c.statesInformer.GetAllPods() {
+		pod := podMeta.Pod
+		if apiext.GetPodQoSClass(pod) == apiext.QoSBE {
+			requestSum = util.GetPodBEMilliCPURequest(pod)
```

Comment:
Here is a typo makes `requestSum +=` as `requestSum =`. And please note `util.GetPodBEMilliCPURequest()` can return `-1` when pod does not specify BatchCPU

Reply:
my bad, fixed.